### DISCUSSION TITLE
"New" options in `visdat` package

### DISCUSSION THREAD

**Discussion by lilyclements** [IDEMSInternational/R-Instat#9769] [Climatic]
I say "new" because I'm not sure how new they are. But some nice neat new ways to easily visualise your data frame in the [visdat package](https://github.com/ropensci/visdat) We currently have under 

Prepare > Check Data > Visualise Data

the `vis_dat`, `vis_miss`, and `vis_guess` functions run. There are now more we could add, or at least consider for elsewhere.

**Visualising Correlations**
We can visualise correlations between all variables, or facet by another variable. For example,

`vis_cor(airquality, facet = Month)`
![image](https://github.com/africanmathsinitiative/R-Instat/assets/21180424/11c4550b-a67d-44c1-9b53-ff4b248c2380)

This looks somewhat similar to something to the right hand panel in the `desplot` package that @rdstern has written a little guide on
<img width="381" alt="image" src="https://github.com/africanmathsinitiative/R-Instat/assets/21180424/9d1b3de7-1d1b-449c-a87b-21a35ad07735">

**Compare Data Frames**
Perhaps relevant around duplicating data frames bit, but this is useful to compare two data frames
```
aq_diff <- airquality
aq_diff[1:10, 1:2] <- NA
vis_compare(airquality, aq_diff)
```
![image](https://github.com/africanmathsinitiative/R-Instat/assets/21180424/8950675c-5e4c-422a-a370-ab8098a0adda)

**Expected values**
From their guide - 
"vis_expect visualises certain conditions or values in your data. For example, If you are not sure whether to expect values greater than 25 in your data (airquality), you could write: vis_expect(airquality, ~.x>=25), and you can see if there are times where the values in your data are greater than or equal to 25:
`vis_expect(airquality, ~.x >= 25)`
![image](https://github.com/africanmathsinitiative/R-Instat/assets/21180424/9648a2a9-8357-4178-b2a5-0513a8de8ad9)
"
I'm not sure where this would be most appropriate in R-Instat. But, perhaps @rdstern can see a use. Either way, useful to mention it as another new option in `visdat`. 

**Visualising Numerical Values**
`vis_value()` is a quick way to visualise just the numerical values in a data frame, on a 0-1 scale. This throws an error if there is a non-numerical variable in the data frame, so if we used it we would want to be careful of that. Not sure again how useful this is, but it shows some structure in the data. Running for just numerical values in the Niger data:

`vis_value(Niger %>% dplyr::select(where(is.numeric)))`

![image](https://github.com/africanmathsinitiative/R-Instat/assets/21180424/db0c2be9-3c15-424f-ad84-464413531a9e)

**Visualising Binary Values**
`vis_binary()` visualises binary values. This doesn't seem to like if we have TRUE/FALSE values. This also throws an error if there is a non-binary variables in the data frame, so if we used it we would want to be careful of that. I can't see too much of the use of this one at the moment.

`vis_binary(dat_bin)`
![image](https://github.com/africanmathsinitiative/R-Instat/assets/21180424/5d4dbc8f-d4b9-485e-855d-7588e7a854b6)

**Comment by rdstern** [comment]
@lilyclements I really like some of these.

<img width="287" alt="image" src="https://github.com/africanmathsinitiative/R-Instat/assets/11226469/0e330cfc-79b3-47a9-9f28-3031268ee70c">

I note that the call to the visdat package is right at the top of our Prepare > Check data dialog and this seems a good place for some of these improvements.  In particular:
a) I really like the option to compare 2 data frames.  We have another compare below, and this feature would be an excellent visual complement.
b) Some of the options only work with columns of a particular type.  We can easily make that resriction in the dialog.

It should be easy and I'll specify the details, so a pull request can be done.  At the same time, I haven't used the 2nd and third buttons much, so we may plan a re-design.  I wonder what has happened with the default size.  I have had to change that quite often.  If is it still a general feature I suggest we make the default larger.

**Comment by lilyclements** [comment]
>  I wonder what has happened with the default size. I have had to change that quite often. If is it still a general feature I suggest we make the default larger.

@rdstern the size is limited to `9e+05` (I assume that was the limit before)

There is also now an option `facet` in all of these functions (and the three we have). I think this is new, but do correct me if I'm wrong. 

E.g. for `vis_miss(Niger, facet = Name)`

![image](https://github.com/africanmathsinitiative/R-Instat/assets/21180424/b31f1dd2-353b-4018-965c-0cedefa3c7d5)

There are two issues with this that I see though: Firstly is that the observations all start on 1, and for those that start later they have a blank at the end (all stations except Niamey Aero)
Secondly is that the % missing is not given if I do with facets. But, I think faceting is still a useful option to see the patterns by a group. Are there more missing values in a certain month for a variable? Etc.

**Comment by rdstern** [comment]
@lilyclemeI really like the facet.  That seems to be a no brainer to me.  In your example above I think Niamey starts in 1940 and the others in 1945, so that's all quite sensible!

**Comment by rdstern** [comment]
@lilyclements could you tempt the Ghana team to work on this a bit.  Not vital, given all the other work, but perhaps one or two improvements could be very easy to include?

**Comment by lilyclements** [comment]
@rdstern a few comments first:

1. I find the name "guess" for the `vis_guess` one quite confusing. Could this be named something else like "Row Type" or "Cell Type"? Or something which explains it a bit more? Let me know what you think. 
**_RDS: Very happy with this change_**
2. Would you consider "Data" being called "Column Type"? I think that Data is quite broad as all options are looking at the data. However, I don't want to confuse the user further by changing it now, or to something they may find less intuitive (just because I find it more intuitive doesn't mean they will).
`RDS: Happy to see this change too.  Maybe Column Types - so with the s?`
3. Do you see the one for checking duplicated data fitting in here, or would we instead want it around the "Duplicated Rows" dialog? I could see an argument for either
**_RDS: Probably with duplicated and leave till later?_**
4. I imagine our `vis_cor` would make more sense around the correlations dialog. I'm not sure if it is checking the data in the same way as our other `vis_` functions are. However, would you say that correlations is a standard way to check data?
**_RDS:  And again left till later._**
This is the current dialog:

<img width="332" alt="image" src="https://github.com/africanmathsinitiative/R-Instat/assets/21180424/f3089eb7-191c-4e11-90a9-1dc2fadf4ac2">

* We have Data, Missing, Guess. We could add in three options "Numerical", "Expect", and "Binary"

**In "Numerical":**
1. We have the `vis_value` function
3. We have a multiple receiver which accepts numerical variables only (we do not have the option for the whole data set)
4. I'll give the parameters below:

| Option | Label name | R Parameter | Parameter Values | Default |
|----------|-----------|-----------|----------|----------|
| Multiple receiver | Numeric Variables: | `data` | values in the multiple receiver ||
| Label and non-editable drop down | Palette: | `viridis_option` | values in the drop down - `"magma"` `"inferno"`, `"plasma"`, `"viridis"`, `"cividis"`| `"viridis"`
| ucrColour | Colour for Missing Values: | `na_colour` | the usual options in ucrColour | `"grey90"` |


**In "BInary":**
1. We have the `vis_binary` function
2. We have a multiple receiver which accepts binary variables only (we do not have the option for the whole data set)
3. I'll give the parameters below:

| Option | Label name | R Parameter | Parameter Values | Default |
|----------|-----------|-----------|----------|----------|
| Multiple receiver | Binary Variables: | `data` | values in the multiple receiver ||
| ucrColour | Colour for Zeros: | `col_zero` | the usual options in ucrColour | `"salmon"` |
| ucrColour | Colour for Ones: | `col_one` | the usual options in ucrColour | `"steelblue2"` |
| ucrColour | Colour for NAs: | `col_na` | the usual options in ucrColour | `"grey90"` |


**In "Expect":**
This will be simple as well, but I am about to go out and don't want to rush so will get back to specifying this one later. 


**We could add to Data and Miss:**
We could offer a checkbox "Facet by" which leads to a receiver in which the user can input a factor variable to facet by. This is  just the parameter `facet` and leads to the variable to facet by.


There are a few simple changes we could implement here. @derekagorhom perhaps this could be a good task for Raymond, or perhaps it's a tall ask! Let me know what you think

**Comment by rdstern** [comment]
@lilyclements I have added to your top comments above.  
I am also looking for quick wins, so starting by items above that would be quick and easy and useful.  They partially fulfil this issue, and can be merged.  Then there continues to be work available, possibly later, on remaining parts.

a) So I wonder about starting with the name changes above and also adding facets?
b) If that is quick, then the numerical option sounds useful, and possibly also quite quick?  It could include facets too perhaps?

What do you think?

**Comment by lilyclements** [comment]
@rdstern sounds good to me. @derekagorhom is this something you could look at with Raymond?

**Comment by Raymond-Andam** [comment]
@rdstern @lilyclements Derrick and I have started working on the issue
